test/cpp_extensions/open_registration_extension/pytorch_openreg/_device_daemon.py

```diff
@@ -75,94 +68,131 @@ def tensor_from_meta(self, meta):
-        empty_res = object()
-        allocator = Allocator()
+    registry = {}
+
+    @register(registry)
+    def deviceCount(self, *args):
+        assert len(args) == 0
+        return self.num_devices
+
+    @register(registry)
+    def getDevice(self):
+        return self.curr_device_idx
+
+    @register(registry)
+    def uncheckedSetDevice(self, *args):
+        assert len(args) == 1
+        self.curr_device_idx = int(args[0])
+
+    @register(registry)
+    def exchangeDevice(self, *args):
+        assert len(args) == 1
+        res = self.curr_device_idx
+        self.curr_device_idx = int(args[0])
+        return res
+
 
+class _Executor:
+    def __init__(self):
+        self.allocator = Allocator()
+
+    def run_forever(self, req_queue, ans_queue):
         # Serve all requests
         while True:
             cmd, *args = req_queue.get()
             log.info("Worker executing: %s", cmd)
-            res = empty_res
-            if cmd == "deviceCount":
-                assert len(args) == 0
-                res = NUM_DEVICES
-            elif cmd == "getDevice":
-                res = CURR_DEVICE_IDX
-            elif cmd == "uncheckedSetDevice":
-                assert len(args) == 1
-                CURR_DEVICE_IDX = int(args[0])
-                res = None
-            elif cmd == "exchangeDevice":
-                assert len(args) == 1
-                res = CURR_DEVICE_IDX
-                CURR_DEVICE_IDX = int(args[0])
-            elif cmd == "malloc":
-                res = allocator.malloc(*args)
-            elif cmd == "free":
-                res = allocator.free(*args)
-            elif cmd == "run_op":
-                op_name, args, kwargs = args
-                run_op(allocator, op_name, args, kwargs)
-                res = None
-            elif cmd == "send_data":
-                assert len(args) == 1
-                res = OpenRegTensorData.from_meta(allocator, args[0])
-            elif cmd == "recv_data":
-                assert len(args) == 2
-                host_tensor, dev_mem = args
-                dev_tensor = OpenRegTensorData.from_meta(allocator, dev_mem)
-                dev_tensor.copy_(host_tensor)
-                res = None
-            elif cmd == "get_op_output_shape":
-                op_name, args, kwargs = args
-                res = run_op(allocator, op_name, args, kwargs).size()
+            if cmd in _Executor.registry:
+                res = _Executor.registry[cmd](self, *args)
             else:
                 log.warning("Bad command in worker")
                 res = "ERROR"
 
-            if res == empty_res:
-                raise RuntimeError("Bad impl didn't return anything")
             log.info("Worker answering to: %s", cmd)
             ans_queue.put(res)
 
+    registry = {}
+
+    @register(registry)
+    def malloc(self, size):
+        return self.allocator.malloc(size)
+
+    @register(registry)
+    def free(self, ptr):
+        return self.allocator.free(ptr)
```

Comment:
This one is interesting.
While we practically need the memory allocated here for running the op, to some extend, in a realistic use case, the allocator would be handled on the driver side right?

I don't think we should change much here except maybe rename this to not say "allocator" and leave a small comment here.
The fancy version might be to have something like a caching allocator where we have a driver side caching allocator that is fast and it call the device malloc/free (that are expensive) only when needed. We shouldn't do that in this PR though.

Reply:
> I don't think we should change much here except maybe rename this to not say "allocator" and leave a small comment here.

I agree that it would be better to rename "allocator" here to distinguish it from OpenRegAllocator.

> The fancy version might be to have something like a caching allocator where we have a driver side caching allocator that is fast and it call the device malloc/free (that are expensive) only when needed.

Sure we can do some caching strategies like CUDACachingAllocator (of course not that complicated), after we complete the basic functions mentioned in Next steps :)